### PR TITLE
Dereddening for standards

### PR DESCRIPTION
**Pull request by karllark**
Add the capability to deredden star data based on FM90+CCM89 parameters.  This supports Milky Way UV extinction curve work where dereddening the standards is needed and has been a common technique for many years (see https://ui.adsabs.harvard.edu/abs/1992AJ....104.1916C/abstract).   For the MW standards, the necessary dereddening info is saved in the dat files.

Also added entry-points (aka system wide aliases) for meplot_spec, meplot_ext, and mecalc_ext that correspond to running the plotting/plot_spec, plotting/plot_ext, and utils/calc_ext scripts.

### CODE REVIEW

**Comment by karllark** [comment]
@mdecleir: when you get time, could you review this PR?  This enables classical MW UV extinction curve work that requires dereddening standards.  I've also cleaned up the scripts a bit and hopefully not introduced to many bugs.  No great rush, just working to push on outstanding PRs before start the next extinction curve project.

**Comment by mdecleir** [comment]
> @mdecleir: when you get time, could you review this PR? This enables classical MW UV extinction curve work that requires dereddening standards. I've also cleaned up the scripts a bit and hopefully not introduced to many bugs. No great rush, just working to push on outstanding PRs before start the next extinction curve project.

Looks like a pretty large PR. I will review it asap. It might have some conflicts with changes I recently made but did not push yet, but let's solve those when they occur.

File: measure_extinction/plotting/plot_spec.py [hunk 1]
```diff
@@ -205,24 +216,33 @@ def plot_multi_spectra(
...
             starobs.data.keys() - (exclude + exclude2)
         )
+        if wavenum:
+            x = 1 / waves
+        else:
+            x = waves
         if range is not None:
-            waves = waves[waves >= range[0]]
-        min_wave = waves[0]
+            x = x[x >= range[0]]
```
Comment (mdecleir) [inline measure_extinction/plotting/plot_spec.py:224]: This will, of course, only work if range is also in wavenumbers. The information about "range" in this function (and others) is currently: Wavelength range to be plotted (in micron) - [min,max]. This should probably be changed to something more general.

Comment (mdecleir) [reply measure_extinction/plotting/plot_spec.py:224]: Ah, I see now that you changed it for the plot_spectrum function below, but not for this one.

File: measure_extinction/stardata.py [hunk 2]
```diff
@@ -1000,6 +1028,56 @@ def _parse_dfile_line(line):
...
+                        "F99_method",
+                    )
+                    self.data[curtype].fluxes[gvals] /= 10 ** (
+                        -0.4 * alav * self.deredden_params["AV"]
+                    )
+                else:
+                    warnings.warn(f"{curtype} cannot be dereddened", UserWarning)
+        else:
+            warnings.warn(
+                "cannont deredden as no dereddening parameters set", UserWarning
```
Comment (mdecleir) [inline measure_extinction/stardata.py:1078]:
```suggestion
                "cannot deredden as no dereddening parameters set", UserWarning
```

File: setup.cfg [hunk 3]
```diff
@@ -44,6 +44,12 @@ norecursedirs =
     measure_extinction/docs/_build/*
     measure_extinction/docs/api/*
 
+[options.entry_points]
+console_scripts =
+    meplot_spec = measure_extinction.plotting.plot_spec:main
+    meplot_ext = measure_extinction.plotting.plot_ext:main
+    mecalc_ext = measure_extinction.utils.calc_ext:main
```
Comment (mdecleir) [inline setup.cfg:51]: Just curious, what exactly is the purpose of this?

**Review by mdecleir — APPROVED** [review]
Nice additions! It looks like these changes should not break anything.
However, it will result in different plots, so I might not pull these changes locally until after the paper is done ;)

File: measure_extinction/plotting/plot_ext.py [hunk 4]
```diff
@@ -443,7 +447,10 @@ def plot_multi_extinction(
     plt.rc("axes.formatter", min_exponent=2)
 
     # create the plot
-    fig, ax = plt.subplots(figsize=(15, len(starpair_list) * 1.25))
+    ysize = 6
+    if spread:
+        ysize += len(starpair_list) * 1.25
+    fig, ax = plt.subplots(figsize=(8, ysize))
```
[the remark this answers is not in the document]
Comment (karllark) [reply measure_extinction/plotting/plot_ext.py:453]: Yeah.  It is clear that the plot size/ratio depends on the wavelength range being plotted.  When I tried to stack UV extinction curves, the figure was very oblong and not very useful.  I realized that this would muck up your work, but don't have any idea how to pick good defaults.  Maybe we should setup a set of data from different wavelengths for such testing?

File: measure_extinction/plotting/plot_ext.py [hunk 5]
```diff
@@ -535,12 +544,18 @@ def plot_multi_extinction(
...
+        xlab = r"$1/\lambda$ [$\mu m^{-1}$]"
+    else:
+        xlab = r"$\lambda$ [$\mu m$]"
+    plt.xlabel(xlab, fontsize=1.5 * fontsize)
     ylabel = extdata._get_ext_ytitle(ytype=extdata.type)
     if spread:
         ylabel += " + offset"
     ax.set_ylabel(ylabel, fontsize=1.5 * fontsize)
 
+    fig.tight_layout()
```
[the remark this answers is not in the document]
Comment (karllark) [reply measure_extinction/plotting/plot_ext.py:557]: My understanding is that the two commands are not the same.  `tight_layout` adjusts plot components and not just whitespace.  But I'm a novice with these commands and have just mainly used `tight_layout` as it seems to work well is almost all cases.   I wan't even aware of the `bbox_inches` option.

File: setup.cfg (hunk 3 above)
Comment (karllark) [reply setup.cfg:51]: These mean that if you have pip installed this package, then you can use `meplot_spec` etc. on the command line from any directory.  Just easier to call the command line scripts then having to remember their full paths.

File: measure_extinction/plotting/plot_spec.py (hunk 1 above)
Comment (karllark) [reply measure_extinction/plotting/plot_spec.py:224]: Ok.  Looked into this and am not not sure what needs fixing.

File: measure_extinction/stardata.py (hunk 2 above)
Comment (karllark) [reply measure_extinction/stardata.py:1078]: Fixed.

File: measure_extinction/plotting/plot_ext.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (karllark) [reply measure_extinction/plotting/plot_ext.py:557]: And this will work for plotting it to the screen too.

File: measure_extinction/plotting/plot_ext.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (mdecleir) [reply measure_extinction/plotting/plot_ext.py:453]: Thinking about it, I might actually already have a solution for this problem. In general, all the plotting settings will differ dependent on wavelength range, type of figure you want to make for your specific purpose, and even personal taste. I wanted to change settings (such as axis limits, tick labels, fontsize, etc.) for every plot in my paper, but it was too difficult to make every setting an optional keyword in the function. So, I just set the defaults in `plot_ext.py` to values that result in reasonably looking plots. And then, I added the line
`return fig, ax` at the end of every plot function (as you can see in the code), such that in my specific plotting scripts, I can further tweak the plots. I now found that I can also change the figure size after the plot was created, using:
`fig.set_figheight(20)` and `fig.set_figwidth(10)`.

So, long story short, we should just pick some default value for all settings and then tweak to our liking in our specific plotting scripts.

File: measure_extinction/plotting/plot_ext.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (mdecleir) [reply measure_extinction/plotting/plot_ext.py:557]: > My understanding is that the two commands are not the same. `tight_layout` adjusts plot components and not just whitespace. But I'm a novice with these commands and have just mainly used `tight_layout` as it seems to work well is almost all cases. I wan't even aware of the `bbox_inches` option.

Yeah, for some reason `tight_layout` seems to make the fontsize smaller as well. I didn't know that. Usually I need bigger fonts to make plots more readable.

File: setup.cfg (hunk 3 above)
Comment (mdecleir) [reply setup.cfg:51]: Cool! Does that work in a script as well, when importing those modules?

File: measure_extinction/plotting/plot_spec.py (hunk 1 above)
Comment (mdecleir) [reply measure_extinction/plotting/plot_spec.py:224]: Everywhere where it says: "Wavelength range to be plotted (in micron) - [min,max]", you should simply delete the "(in micron)". Then it is up to the user to make sure that they provide the ranges to be plotted in the correct units (either micron or 1/micron)... Or maybe it would be better to mention that, i.e.:
"Wavelength range to be plotted (in micron when plotting wavelengths or in 1/micron when plotting wavenumber) - [min,max]"

**Review by mdecleir — APPROVED** [review]

**Comment by mdecleir** [comment]
Hum, I did not mean to approve all changes. I just wanted to add the extra comment about the wavelength ranges...

File: measure_extinction/plotting/plot_ext.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (karllark) [reply measure_extinction/plotting/plot_ext.py:453]: Sounds like a plan.  Thanks for including the code on how to do this - will be very useful!

File: measure_extinction/plotting/plot_ext.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (karllark) [reply measure_extinction/plotting/plot_ext.py:557]: Maybe we need to add a fontsize keyword and then have everything scaled off this keyword?  Not for this PR, but an idea.

File: setup.cfg (hunk 3 above)
Comment (karllark) [reply setup.cfg:51]: Not quite sure what you mean.

Comment (mdecleir) [reply setup.cfg:51]: So, you say you can use those shortcuts on the command line to call the scripts. My question is if you can also do
`import meplot_spec` in another python script, instead of
`import measure_extinction.plotting.plot_spec` e.g.

Comment (karllark) [reply setup.cfg:51]: No.   This only works for the command line.  Imports have to use the full package info.

**Comment by karllark** [comment]
@mdecleir : ok to merge?

**Comment by mdecleir** [comment]
> @mdecleir : ok to merge?

Yes!
